guidance/models/_model.py

```diff
@@ -763,8 +1271,256 @@ def _run_stateless(self, stateless_function, temperature=0.0, top_p=1.0, n=1):
 
         logger.debug("finish Model._run_stateless")
 
+        lm.metrics = lm.engine.metrics.model_copy(deep=True)
+
         return lm
 
+    def get_per_token_stats(self) -> list[GenTokenExtra]:
+        """Get the per token stats (prob, top-k, latency, etc.) for the model.
+
+        Returns
+            List[GenTokenExtra]: a list of GenToken objects.
+        """
+
+        paths = []
+        model = self
+        while model is not None:
+            paths.append(model)
+            if model._parent_id is None:
+                break
+
+            model: "Model" = self.engine.model_dict[model._parent_id]
+        paths.reverse()
+
+        vis_chunks: list[VisBytesChunk] = [
+            path.vis_chunk for path in paths if path.vis_chunk is not None
+        ]
+
+        gen_tokens_infos: list[tuple[int, float, list[GenToken]]] = []
+        gen_tokens_indices: list[int] = []
+        for vis_chunk in vis_chunks:
+            for engine_output in vis_chunk.engine_outputs:
+                gen_tokens_infos.append(
+                    (
+                        engine_output.issued_token.token_id,
+                        engine_output.issued_token.latency_ms,
+                        engine_output.masked_top_k,
+                    )
+                )
+            gen_tokens_indices.append(len(gen_tokens_infos) - 1)
+
+        text = self._state
+        token_ids = self.engine.tokenizer.encode(text.encode("utf-8"))
```

Comment:
Be very careful here.
1. We may need to ensure a bos_token prefixes this, as there is additional work being done in the parser to do this (seems like this and the parser should use a common function)
2. In working through this PR, I have realized that we're not always sending proper tokenizations to the engine in our generate loop. Re-tokenizing like this at the end may therefore give tokens (and probabilities) that mismatch the actual ones that are seen during generation.
3. Both points above additionally mean higher costs than otherwise expected since the KV cache may be invalidated.

The point about improper tokenization is something that we should probably handle lower down, either in llguidance or in the token parser. Need to chat with @Harsha-Nori and/or @mmoskal about this.

Reply:
great catch. Will fix this BOS issue.
For 3) We don't really use KV cache to calculate the final probs as the engine will not generate probs for token cached.

Reply:
@hudson-ai added a check to append BOS token into token_ids before calling get_per_token_topk_probs.